crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,321 @@
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{Context, Poll};
+use std::time::Duration;
+use sui_types::base_types::{SequenceNumber, TransactionDigest};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use tonic::body::BoxBody;
+use tonic::codegen::Service;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Streaming;
+
+const COLUMN_FAMILY_NAME: &str = "sui";
+const COLUMN_QUALIFIER: &str = "";
+const OBJECTS_TABLE: &str = "objects";
+const CHECKPOINTS_TABLE: &str = "checkpoints";
+const TRANSACTIONS_TABLE: &str = "transactions";
+
+#[derive(Clone)]
+struct AuthChannel {
+    channel: Channel,
+    policy: String,
+    token_provider: Option<Arc<dyn TokenProvider>>,
+    token: Arc<RwLock<Option<Arc<Token>>>>,
+}
+
+#[derive(Clone)]
+pub struct BigTableClient {
+    table_prefix: String,
+    client: BigtableInternalClient<AuthChannel>,
+}
+
+#[async_trait]
+impl KeyValueStore for BigTableClient {
+    async fn get_objects(&mut self, object_keys: &[ObjectKey]) -> Result<Vec<Object>> {
+        let keys: Result<_, _> = object_keys.iter().map(|key| bcs::to_bytes(&key)).collect();
+        self.multi_get(OBJECTS_TABLE, keys?).await
+    }
+
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>> {
+        let keys = transactions.iter().map(|tx| tx.inner().to_vec()).collect();
+        self.multi_get(TRANSACTIONS_TABLE, keys).await
+    }
+
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>> {
+        let key = sequence_number.value().to_be_bytes().to_vec();
+        let mut result = self.multi_get(CHECKPOINTS_TABLE, vec![key]).await?;
+        Ok(result.pop())
+    }
+
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()> {
+        let mut items = Vec::with_capacity(objects.len());
+        for object in objects {
+            let object_key = ObjectKey(object.id(), object.version());
+            items.push((bcs::to_bytes(&object_key)?, object));
+        }
+        self.multi_set(OBJECTS_TABLE, items).await
+    }
+
+    async fn save_transactions(&mut self, transactions: &[TransactionData]) -> Result<()> {
+        let items = transactions.iter().map(|transaction| {
+            (
+                transaction.transaction.digest().inner().to_vec(),
+                transaction,
+            )
+        });
+        self.multi_set(TRANSACTIONS_TABLE, items).await
+    }
+
+    async fn save_checkpoint(&mut self, checkpoint: &CheckpointData) -> Result<()> {
+        let summary = &checkpoint.checkpoint_summary;
+        let contents = &checkpoint.checkpoint_contents;
+        let key = summary.sequence_number.to_be_bytes().to_vec();
```

Comment:
Key as bigendian bytes is perfect for the key. I do wonder how or if this will lead to a hot writer since all writes will be sequential (maybe something to look into?)

Same for txn i think we should write the checkpoint into multiple columns, probably (summary, validator signature, contents)

Reply:
I don't think it will be an issue for the writer. We have a single one and it's only a couple of checkpoints per second. But might be a bit more relevant concern for the reader